File: MiniDOMBase/DMD/Anagrafica/Persone/Info/CPersonaInfo.cs
using DMD.XML;
using System;
using DMD;
using DMD.Databases;
using DMD.Databases.Collections;
using static minidom.Sistema;
using static minidom.Anagrafica;


namespace minidom
{
    public partial class Anagrafica
    {

        /// <summary>
        /// Oggetto "light" che racchiude informazioni su una persona o un'azienda
        /// </summary>
        [Serializable]
        public class CPersonaInfo 
            : IDMDXMLSerializable, IComparable, IComparable<CPersonaInfo>
        {
            private CKeyCollection m_Attributes;
            private string m_IconURL;
            private bool m_Deceduto;
            private int m_IDPersona;
            [NonSerialized] private CPersona m_Persona;
            private string m_NomePersona;
            private string m_Notes;

            /// <summary>
            /// Costruttore
            /// </summary>
            public CPersonaInfo()
            {
                ////DMDObject.IncreaseCounter(this);
                m_IDPersona = 0;
                m_Persona = null;
                m_NomePersona = "";
                m_Notes = "";
                m_IconURL = "";
                m_Deceduto = false;
                this.m_Attributes = new CKeyCollection();
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="persona"></param>
            public CPersonaInfo(CPersona persona) : this()
            {
                if (persona is null)
                    throw new ArgumentNullException("persona");
                Parse(persona);
            }

            /// <summary>
            /// Inizializza l'oggetto con la persona e prepara le info
            /// </summary>
            /// <param name="persona"></param>
            protected virtual void Parse(CPersona persona)
            {
                if (persona is null)
                    throw new ArgumentNullException("persona");

                m_Persona = persona;
                m_IconURL = persona.IconURL;
                m_Deceduto = persona.Deceduto;
                m_IDPersona = DBUtils.GetID(persona, 0);
                m_NomePersona = persona.Nominativo;
                m_Notes = "";
                string txtN = Strings.EQ(m_Persona.Sesso , "F", true)? "Nata" : "Nato";
                // If TypeOf (Me.m_Persona) Is CPersonaFisica Then
                // If Me.m_Persona.ImpiegoPrincipale IsNot Nothing Then
                // With DirectCast(Me.m_Persona, CPersonaFisica).ImpiegoPrincipale.ToString <> ""
                // If 
                // End With

                // End If
                // End If
                if (!string.IsNullOrEmpty(m_Persona.NatoA.ToString()))
                    m_Notes = DMD.Strings.Combine(m_Notes, txtN + " a: " + m_Persona.NatoA.ToString(), ", ");
                if (m_Persona.DataNascita.HasValue)
                    m_Notes = DMD.Strings.Combine(m_Notes, txtN + " il: " + Sistema.Formats.FormatUserDate(m_Persona.DataNascita), ", ");
                if (!string.IsNullOrEmpty(m_Persona.DomiciliatoA.ToString()))
                {
                    // Me.m_Notes = DMD.Strings.Combine(Me.m_Notes, "Domiciliato a: " & Me.m_Persona.DomiciliatoA.ToString, ", ")
                    m_Notes = DMD.Strings.Combine(m_Notes, m_Persona.DomiciliatoA.ToString(), ", ");
                }
                else if (!string.IsNullOrEmpty(m_Persona.ResidenteA.ToString()))
                {
                    // Me.m_Notes = DMD.Strings.Combine(Me.m_Notes, "Residente a: " & Me.m_Persona.ResidenteA.ToString, ", ")
                    m_Notes = DMD.Strings.Combine(m_Notes, m_Persona.ResidenteA.ToString(), ", ");
                }

                if (!string.IsNullOrEmpty(m_Persona.CodiceFiscale))
                    m_Notes = DMD.Strings.Combine(m_Notes, "C.F.: " + m_Persona.CodiceFiscale, ", ");
                Attributes.Add("Titolo", m_Persona.Titolo);
            }

            void IDMDXMLSerializable.XMLSerialize(XMLWriter writer)
            {
                this.XMLSerialize(writer);
            }

            void IDMDXMLSerializable.SetFieldInternal(string fieldName, object fieldValue)
            {
                this.SetFieldInternal(fieldName, fieldValue);
            }

            /// <summary>
            /// Attributi aggiuntivi
            /// </summary>
            public CKeyCollection Attributes
            {
                get
                {
                    return m_Attributes;
                }
            }

            /// <summary>
            /// Icona della persona
            /// </summary>
            public string IconURL
            {
                get
                {
                    return m_IconURL;
                }

                set
                {
                    m_IconURL = DMD.Strings.Trim(value);
                }
            }

            /// <summary>
            /// Se vero indica che la persona è deceduta
            /// </summary>
            public bool Deceduto
            {
                get
                {
                    return m_Deceduto;
                }

                set
                {
                    m_Deceduto = value;
                }
            }

            /// <summary>
            /// ID della persona
            /// </summary>
            public int IDPersona
            {
                get
                {
                    return DBUtils.GetID(m_Persona, m_IDPersona);
                }

                set
                {
                    if (IDPersona == value)
                        return;
                    m_IDPersona = value;
                    m_Persona = null;
                }
            }

            /// <summary>
            /// Persona
            /// </summary>
            public CPersona Persona
            {
                get
                {
                    if (m_Persona is null)
                        m_Persona = Persone.GetItemById(m_IDPersona);
                    return m_Persona;
                }

                set
                {
                    m_Persona = value;
                    m_IDPersona = DBUtils.GetID(value, 0);
                    if (value is object)
                        m_NomePersona = value.Nominativo;
                }
            }

            /// <summary>
            /// Nominativo della persona
            /// </summary>
            public string NomePersona
            {
                get
                {
                    return m_NomePersona;
                }

                set
                {
                    m_NomePersona = DMD.Strings.Trim(value);
                }
            }

            /// <summary>
            /// Descrizione
            /// </summary>
            public string Notes
            {
                get
                {
                    return m_Notes;
                }

                set
                {
                    m_Notes = value;
                }
            }

            /// <summary>
            /// Serializzazione xml
            /// </summary>
            /// <param name="writer"></param>
            protected virtual void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("IDPersona", m_IDPersona);
                writer.WriteAttribute("NomePersona", m_NomePersona);
                writer.WriteAttribute("Deceduto", m_Deceduto);
                writer.WriteAttribute("IconURL", m_IconURL);
                writer.WriteTag("Attributes", Attributes);
                writer.WriteTag("Notes", m_Notes);
            }

            /// <summary>
            /// Deserializzazione xml
            /// </summary>
            /// <param name="fieldName"></param>
            /// <param name="fieldValue"></param>
            protected virtual void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "IDPersona":
                        {
                            m_IDPersona = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "NomePersona":
                        {
                            m_NomePersona = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Notes":
                        {
                            m_Notes = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Deceduto":
                        {
                            m_Deceduto = DMD.XML.Utils.Serializer.DeserializeBoolean(fieldValue) == true;
                            break;
                        }

                    case "IconURL":
                        {
                            m_IconURL = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Attributes":
                        {
                            this.m_Attributes = (CKeyCollection)fieldValue;
                            break;
                        }
                }
            }

            /// <summary>
            /// Compara due oggetti
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public int CompareTo(CPersonaInfo obj)
            {
                return DMD.Strings.Compare(NomePersona, obj.NomePersona, true);
            }

            int IComparable.CompareTo(object obj)
            {
                return CompareTo((CPersonaInfo)obj);
            }

            /// <summary>
            /// Distruttore
            /// </summary>
            ~CPersonaInfo()
            {
               // //DMDObject.DecreaseCounter(this);
            }
        }
    }
}

File: MiniDOMBase/DMD/Anagrafica/Persone/CPersonaInfo.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.Databases



Partial Public Class Anagrafica

    <Serializable> _
    Public Class CPersonaInfo
        Implements XML.IDMDXMLSerializable
        Implements IComparable

        Private m_Attributes As New CKeyCollection(Of String)
        Private m_IconURL As String
        Private m_Deceduto As Boolean
        Private m_IDPersona As Integer
        <NonSerialized>
        Private m_Persona As CPersona
        Private m_NomePersona As String
        Private m_Notes As String

        Public Sub New()
            DMDObject.IncreaseCounter(Me)
            Me.m_IDPersona = 0
            Me.m_Persona = Nothing
            Me.m_NomePersona = ""
            Me.m_Notes = ""
            Me.m_IconURL = ""
            Me.m_Deceduto = False
        End Sub

        Public Sub New(ByVal persona As CPersona)
            Me.New()
            If (persona Is Nothing) Then Throw New ArgumentNullException("persona")
            Me.Parse(persona)
        End Sub

        Protected Sub Parse(ByVal persona As CPersona)
            Me.m_IconURL = persona.IconURL
            Me.m_Deceduto = persona.Deceduto
            Me.m_IDPersona = GetID(persona)
            Me.m_Persona = persona
            Me.m_NomePersona = persona.Nominativo
            Me.m_Notes = ""
            Dim txtN As String = IIf(Me.m_Persona.Sesso = "F", "Nata", "Nato")
            'If TypeOf (Me.m_Persona) Is CPersonaFisica Then
            '    If Me.m_Persona.ImpiegoPrincipale IsNot Nothing Then
            '        With DirectCast(Me.m_Persona, CPersonaFisica).ImpiegoPrincipale.ToString <> ""
            '            If 
            '        End With

            '    End If
            'End If
            If (Me.m_Persona.NatoA.ToString <> "") Then Me.m_Notes = Strings.Combine(Me.m_Notes, txtN & " a: " & Me.m_Persona.NatoA.ToString, ", ")
            If (Me.m_Persona.DataNascita.HasValue) Then Me.m_Notes = Strings.Combine(Me.m_Notes, txtN & " il: " & Formats.FormatUserDate(Me.m_Persona.DataNascita), ", ")
            If (Me.m_Persona.DomiciliatoA.ToString <> "") Then
                'Me.m_Notes = Strings.Combine(Me.m_Notes, "Domiciliato a: " & Me.m_Persona.DomiciliatoA.ToString, ", ")
                Me.m_Notes = Strings.Combine(Me.m_Notes, Me.m_Persona.DomiciliatoA.ToString, ", ")
            ElseIf (Me.m_Persona.ResidenteA.ToString <> "") Then
                'Me.m_Notes = Strings.Combine(Me.m_Notes, "Residente a: " & Me.m_Persona.ResidenteA.ToString, ", ")
                Me.m_Notes = Strings.Combine(Me.m_Notes, Me.m_Persona.ResidenteA.ToString, ", ")
            End If
            If (Me.m_Persona.CodiceFiscale <> "") Then Me.m_Notes = Strings.Combine(Me.m_Notes, "C.F.: " & Me.m_Persona.CodiceFiscale, ", ")
            Me.Attributes.Add("Titolo", Me.m_Persona.Titolo)
        End Sub

        Public ReadOnly Property Attributes As CKeyCollection(Of String)
            Get
                Return Me.m_Attributes
            End Get
        End Property

        Public Property IconURL As String
            Get
                Return Me.m_IconURL
            End Get
            Set(value As String)
                Me.m_IconURL = Trim(value)
            End Set
        End Property

        Public Property Deceduto As Boolean
            Get
                Return Me.m_Deceduto
            End Get
            Set(value As Boolean)
                Me.m_Deceduto = value
            End Set
        End Property

        Public Property IDPersona As Integer
            Get
                Return GetID(Me.m_Persona, Me.m_IDPersona)
            End Get
            Set(value As Integer)
                If (Me.IDPersona = value) Then Exit Property
                Me.m_IDPersona = value
                Me.m_Persona = Nothing
            End Set
        End Property

        Public Property Persona As CPersona
            Get
                If (Me.m_Persona Is Nothing) Then Me.m_Persona = Anagrafica.Persone.GetItemById(Me.m_IDPersona)
                Return Me.m_Persona
            End Get
            Set(value As CPersona)
                Me.m_Persona = value
                Me.m_IDPersona = GetID(value)
                If Not (value Is Nothing) Then Me.m_NomePersona = value.Nominativo
            End Set
        End Property

        Public Property NomePersona As String
            Get
                Return Me.m_NomePersona
            End Get
            Set(value As String)
                Me.m_NomePersona = Trim(value)
            End Set
        End Property

        Public Property Notes As String
            Get
                Return Me.m_Notes
            End Get
            Set(value As String)
                Me.m_Notes = value
            End Set
        End Property

        Protected Overridable Sub XMLSerialize(ByVal writer As minidom.XML.XMLWriter) Implements XML.IDMDXMLSerializable.XMLSerialize
            writer.WriteAttribute("IDPersona", Me.m_IDPersona)
            writer.WriteAttribute("NomePersona", Me.m_NomePersona)
            writer.WriteAttribute("Deceduto", Me.m_Deceduto)
            writer.WriteAttribute("IconURL", Me.m_IconURL)
            writer.WriteTag("Notes", Me.m_Notes)
            writer.WriteTag("Attributes", Me.Attributes)
        End Sub

        Protected Overridable Sub SetFieldInternal(ByVal fieldName As String, ByVal fieldValue As Object) Implements XML.IDMDXMLSerializable.SetFieldInternal
            Select Case (fieldName)
                Case "IDPersona" : Me.m_IDPersona = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "NomePersona" : Me.m_NomePersona = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Notes" : Me.m_Notes = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Deceduto" : Me.m_Deceduto = XML.Utils.Serializer.DeserializeBoolean(fieldValue)
                Case "IconURL" : Me.m_IconURL = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Attributes"
                    If (TypeOf (fieldValue) Is CKeyCollection) Then
                        With DirectCast(fieldValue, CKeyCollection)
                            For Each k As String In .Keys
                                Me.m_Attributes.Add(k, .Item(k))
                            Next
                        End With
                    End If
            End Select
        End Sub

        Public Function CompareTo(obj As CPersonaInfo) As Integer
            Return Strings.Compare(Me.NomePersona, obj.NomePersona, CompareMethod.Text)
        End Function

        Private Function CompareTo1(obj As Object) As Integer Implements IComparable.CompareTo
            Return Me.CompareTo(obj)
        End Function

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
            DMDObject.DecreaseCounter(Me)
        End Sub
    End Class






End Class